### DISCUSSION TITLE
what is it meaning?

### DISCUSSION THREAD

**Discussion by zffocussss** [Kong/kong#7952] [Help] [answered]
Hi team:
   I saw some examples in customer plugins. who can tell me what is it meaning?
I guess it is to get the plugin name from a directory name.
but I can not figure out it.

```
-- Grab pluginname from module name
local plugin_name = ({...})[1]:match("^kong%.plugins%.([^%.]+)")

```

**Comment by kikito** [comment, marked as answer]
Hello,

That Lua code can be divided into two parts.

The first part is `({...})[1]` and it can be ... "misterious".

The part that makes it misterious is a piece of Lua knowledge that is somewhat obscure: The `...` varargs variable can be used outside of functions. When used that way, the first value of the list is the *string used to require the module*.

In other words, if you have a module that is `foo/bar/baz.lua`, and is required by `require "foo.bar.baz"`, then the `...` variable in that module will hold `"foo.bar.baz"`.

```
-- foo/bar/baz.lua
print(...)
```

```
-- main.lua
require "foo.bar.baz" -- will print "foo.bar.baz"
```

The `({})[1]` around the `...` is just a somewhat convoluted way to get the first item of the `...` list. In Lua there is a simpler way to do this, by using just parenthesis. So instead of `({...})[1]` the first part could be only `(...)`.

The second part is using the string.match function with [string patterns](https://www.lua.org/pil/20.2.html) and a [capture](https://www.lua.org/pil/20.3.html) (`([^%.]+)`) at the end. It means "take the previous string" make sure that it begins with `kong.plugins.` and then take the rest and return it.

The end result is that the variable `plugin_name` ends up containing the plugin name. For example on a plugin required via `require "kong.plugins.colored-logs"` the variable will contain the string `"colored-logs"`

**Reply by zffocussss** [reply]
thanks ,you are professional
